fix(ast_macros): raise compile error on invalid `generate_derive` input.

rzvxa opened the pull request:
It checks 2 things. 1) The input is a supported derive 2) The given identifier is the same as the fully qualified target trait.

The latter makes sure that the trait for derive is included in the scope.

Part of oxc-project/backlog#124

Here's an expanded example of how we assert traits:

```rust
const _:() = {
    {
        trait AssertionTrait: ::oxc_allocator::CloneIn<'static> {}
        impl<T: CloneIn<'static>> AssertionTrait for T {}
    };
};
```

It makes sure `CloneIn` is the same as `::oxc_allocator::CloneIn` and more importantly requires the user to include the trait if they wish to use it with `generate_derive`.

It also provides LSP jump to definition.

rzvxa commented on crates/oxc_ast_macros/src/lib.rs:
```diff
@@ -1,84 +1,127 @@
-use proc_macro::TokenStream;
```
It shouldn't be I have my git configured to replace CRLF with LF on staging. I'll force a commit to fix it, Not sure why this happened.

rzvxa replied:
In the meantime you can see the gist of this PR in my updated comment above.

overlookmotel replied:
Yes, that is indeed it. I'll push a PR to fix the line breaks once #4747 is merged, and that'll make the diff for this PR clearer.

rzvxa replied:
I'll appreciate it. If it is inconvenient for you I can also fix it, I have a Linux machine and there is also virtualization and WSL.

overlookmotel commented:
This is absolutely ingenious!

You can fool it with:

```rs
use oxc_span::GetSpan as GetSpanReal;
trait GetSpan: GetSpanReal {}

#[ast]
#[generate_derive(GetSpan)]
struct Foo { /* blah */ }
```

But I can't see any way to catch that one, and anyone who does such a bizarre thing probably deserves everything they get!

I've pushed 2 more commits.

1. Adding comments. I've removed the stuff about "all computation is cached... without any complex operation" because I don't think we can reasonably claim there's no complexity any more.
2. A nit - just removing an unnecessary `;` from generated code.

If you're happy with those changes, please go ahead and merge.

rzvxa commented:
> You can fool it with:
> 
> ```rust
> use oxc_span::GetSpan as GetSpanReal;
> trait GetSpan: GetSpanReal {}
> 
> #[ast]
> #[generate_derive(GetSpan)]
> struct Foo { /* blah */ }
> ```
> 
> But I can't see any way to catch that one, and anyone who does such a bizarre thing probably deserves everything they get!

Crap, You are right; And I can't figure out how exactly we can prevent it. If I can think of one I'll follow it up with another PR.

rzvxa committed:
fix(ast_macros): raise compile error on invalid `generate_derive` input. (#4766)

It checks 2 things. 1) The input is a supported derive 2) The given identifier is the same as the fully qualified target trait.

The latter makes sure that the trait for derive is included in the scope.

Part of #4704

Here's an expanded example of how we assert traits:

```rust
const _:() = {
    {
        trait AssertionTrait: ::oxc_allocator::CloneIn<'static> {}
        impl<T: CloneIn<'static>> AssertionTrait for T {}
    };
};
```

It makes sure `CloneIn` is the same as `::oxc_allocator::CloneIn` and more importantly requires the user to include the trait if they wish to use it with `generate_derive`.

It also provides LSP jump to definition.

overlookmotel commented:
> Crap, You are right; And I can't figure out how exactly we can prevent it. If I can think of one I'll follow it up with another PR.

Just to say, this is a tiny point. For all practical purposes, this works. And I noticed this morning that "jump to definition" now works too, which is brilliant.

rzvxa commented:
> > Crap, You are right; And I can't figure out how exactly we can prevent it. If I can think of one I'll follow it up with another PR.
> 
> Just to say, this is a tiny point. For all practical purposes, this works. And I noticed this morning that "jump to definition" now works too, which is brilliant.

Yeah, the jump to definition is a nice plus.
